### DISCUSSION TITLE
How to pre-resolve dependencies into `pyodide-lock.json` without pyodide?

### DISCUSSION THREAD

**Discussion by CNSeniorious000** [pyodide/pyodide#4793] [Q&A]
If I install a package using micropip, there will be a recursion to resolve every dependency, which is a waste of time.

So I want to resolve and lock the versions at build time to flatten the dependencies, which enables me to install them when calling `loadPyodide` as well.

Currently a possible way is to use `micropip.freeze()` after installing the first time, but this requires running micropip inside pyodide runtime, which is hard to do it in CI. But dor now I think it is the only way to get the right lock file.

I know there is `pyodide-lock` to manipulate `pyodide-lock.json`, but it seems to require whl files instead of the requirement itself. So will there be an API in `micropip` or `pyodide-lock` to resolve dependencies for pyodide environment in non-pyodide environments?

I mean, like `pyodide lock add a>=1.0 b~=2.3.4` to resolve urls for their wheels and lock them into the `pyodide-lock.json`? It would be great to be able to directly add from a `requirements.txt` or a `pyproject.toml` too.

**Comment by hoodmane** [comment]
Hi @CNSeniorious000. You have several times now asked about exactly what I am working on or thinking about. Can you contact me off of github and can we set up a call? We should be in closer contact. You almost certainly have a bunch of ideas that I haven't had yet.

On the specific topic, see:
https://github.com/pyodide/pyodide/pull/4773
https://github.com/pyodide/micropip/issues/107
https://github.com/pyodide/micropip/pull/109

**Reply by hoodmane** [reply]
CC @jaraco

**Reply by CNSeniorious000** [reply]
Thank you for providing those references. I now followed the current progress on this topic.
